File: backend/app/models/quantum.py
from typing import Literal

from pydantic import BaseModel, Field


GateName = Literal[
    "H",
    "X",
    "Y",
    "Z",
    "S",
    "T",
    "CX",
    "CNOT",
    "M",
]


class GateOperation(BaseModel):
    gate: GateName

    qubit: int | None = None

    control: int | None = None

    target: int | None = None

    column: int | None = None


class SimulationRequest(BaseModel):
    num_qubits: int = Field(
        default=2,
        ge=1,
        le=12,
    )

    gates: list[GateOperation] = []

    shots: int = Field(
        default=1024,
        ge=1,
        le=100000,
    )


class SimulationResponse(BaseModel):
    num_qubits: int
    circuit: list[dict]
    statevector: list[dict[str, float]]
    probabilities: dict[str, float]
    counts: dict[str, int]
    shots: int

File: frontend/lib/quantum.ts
export const SINGLE_QUBIT_GATES = [
  "H",
  "X",
  "Y",
  "Z",
  "S",
  "T",
] as const;

export type SingleQubitGate =
  (typeof SINGLE_QUBIT_GATES)[number];

export type GateType =
  | SingleQubitGate
  | "CNOT"
  | "M";

export interface SingleQubitOperation {
  id: string;
  type: SingleQubitGate | "M";
  qubit: number;
  column: number;
}

export interface CNOTOperation {
  id: string;
  type: "CNOT";
  control: number;
  target: number;
  column: number;
}

export type CircuitOperation =
  | SingleQubitOperation
  | CNOTOperation;

export interface QuantumCircuit {
  numQubits: number;
  numColumns: number;
  operations: CircuitOperation[];
}

export interface SimulationResult {
  num_qubits: number;

  circuit: Record<string, unknown>[];

  statevector: {
    real: number;
    imaginary: number;
    magnitude: number;
    phase: number;
  }[];

  probabilities: Record<string, number>;

  counts: Record<string, number>;

  bloch_vectors: BlochVector[];

  shots: number;
}

export interface BlochVector {
  qubit: number;
  x: number;
  y: number;
  z: number;
}